File: backend/app/services/profile_service.py
"""Student & recruiter profile management + strength computation."""
from __future__ import annotations

from app.core.exceptions import NotFoundError
from app.models.user import StudentProfile
from app.repositories.application_repo import ApplicationRepository
from app.repositories.misc_repo import ResumeRepository
from app.repositories.user_repo import (
    RecruiterProfileRepository,
    StudentProfileRepository,
    UserRepository,
)
from app.schemas.company import OnboardingRequest, StudentProfileUpdate
from app.services import ai_engine


class ProfileService:
    def __init__(
        self,
        students: StudentProfileRepository,
        recruiters: RecruiterProfileRepository,
        users: UserRepository,
        resumes: ResumeRepository,
        applications: ApplicationRepository,
    ) -> None:
        self.students = students
        self.recruiters = recruiters
        self.users = users
        self.resumes = resumes
        self.applications = applications

    async def get_student(self, user_id: str) -> StudentProfile:
        profile = await self.students.get_by_user(user_id)
        if not profile:
            raise NotFoundError("Student profile not found.")
        return profile

    async def _recompute(self, profile: StudentProfile) -> StudentProfile:
        data = profile.model_dump()
        strength = ai_engine.profile_strength(data)
        apps = await self.applications.count({"student_id": profile.user_id})
        resume_ats = 0
        if profile.active_resume_id:
            resume = await self.resumes.get(profile.active_resume_id)
            if resume:
                resume_ats = resume.analysis.ats_score
        readiness = ai_engine.placement_readiness(
            data, applications=apps, resume_ats=resume_ats
        )
        updated = await self.students.update(
            str(profile.id),
            {"profile_strength": strength, "placement_readiness": readiness},
        )
        return updated or profile

    async def update_student(self, user_id: str, data: StudentProfileUpdate) -> StudentProfile:
        profile = await self.get_student(user_id)
        changes = data.model_dump(exclude_none=True)
        if "skills" in changes:
            changes["skills"] = [s.lower().strip() for s in changes["skills"]]
        updated = await self.students.update(str(profile.id), changes)
        return await self._recompute(updated or profile)

    async def complete_onboarding(self, user_id: str, data: OnboardingRequest) -> StudentProfile:
        profile = await self.get_student(user_id)
        await self.students.update(
            str(profile.id),
            {
                "headline": data.headline,
                "location": data.location,
                "skills": [s.lower().strip() for s in data.skills],
                "education": [e.model_dump() for e in data.education],
                "preferences": data.preferences.model_dump(),
            },
        )
        await self.users.update(user_id, {"onboarding_completed": True})
        refreshed = await self.get_student(user_id)
        return await self._recompute(refreshed)

File: backend/src/services/profileService.js
"use strict";

/** Student & recruiter profile management + strength computation.
 * Mirrors app/services/profile_service.py. */

const { StudentProfile, Resume, Application, User } = require("../models");
const aiEngine = require("../utils/aiEngine");
const { NotFoundError } = require("../utils/errors");

async function getStudent(userId) {
  const profile = await StudentProfile.findOne({ user_id: userId });
  if (!profile) throw new NotFoundError("Student profile not found.");
  return profile;
}

async function recompute(profile) {
  const data = profile.toObject({ flattenMaps: true });
  const strength = aiEngine.profileStrength(data);
  const apps = await Application.countDocuments({ student_id: profile.user_id });
  let resumeAts = 0;
  if (profile.active_resume_id) {
    const resume = await Resume.findById(profile.active_resume_id);
    if (resume) resumeAts = resume.analysis.ats_score;
  }
  const readiness = aiEngine.placementReadiness(data, { applications: apps, resumeAts });
  profile.profile_strength = strength;
  profile.placement_readiness = readiness;
  await profile.save();
  return profile;
}

async function updateStudent(userId, data) {
  const profile = await getStudent(userId);
  const changes = { ...data };
  if (Array.isArray(changes.skills)) {
    changes.skills = changes.skills.map((s) => s.toLowerCase().trim());
  }
  profile.set(changes);
  await profile.save();
  return recompute(profile);
}

async function completeOnboarding(userId, data) {
  const profile = await getStudent(userId);
  profile.set({
    headline: data.headline,
    location: data.location,
    skills: data.skills.map((s) => s.toLowerCase().trim()),
    education: data.education || [],
    preferences: data.preferences || { roles: [], locations: [], job_types: [] },
  });
  await profile.save();
  await User.updateOne({ _id: userId }, { $set: { onboarding_completed: true } });
  return recompute(profile);
}

module.exports = { getStudent, updateStudent, completeOnboarding, recompute };
